discord_bot2.py

```diff
@@ -67,6 +68,152 @@ def from_json(jobj: dict, thread: discord.Thread) -> 'Conversation':
+
+
+async def restart(message):
+    """
+    Restart the bot
+    :param message: The message that triggered the restart
+    """
+    await message.channel.send(f'Restart requested.')
+    await execute_command('git fetch', message.channel)
+    await execute_command('git reset --hard', message.channel)
+    await execute_command('git clean -f', message.channel)
+    await execute_command('git pull --rebase=false', message.channel)
+    await execute_command('poetry install', message.channel)
+    await message.channel.send(f'Restarting.')
+    subprocess.Popen(["bash", "restart.sh"])
+    return
+
+
+async def control_on_message(message):
+    """
+    This function is called whenever the bot sees a message in a control channel
+    :param message:
+    :return:
+    """
+    content = message.content
+    if content.startswith('!restart'):
+        await restart(message)
+
+    elif content.startswith('!log'):
+        await message.channel.send(file=discord.File('/tmp/duck.log'))
+
+    elif content.startswith('!rmlog'):
+        await execute_command("rm /tmp/duck.log", message.channel)
+        await execute_command("touch /tmp/duck.log", message.channel)
+
+    elif content.startswith('!status'):
+        await message.channel.send('I am alive.')
+
+    elif content.startswith('!help'):
+        await display_help(message)
+    elif content.startswith('!'):
+        await message.channel.send('Unknown command. Try !help')
+
+
+async def query(conversation: Conversation, message_text: str):
+    """
+    Query the OPENAI API
+    """
+    logging.debug(f"User said: {message_text}")
+
+    conversation.messages.append(dict(role='user', content=message_text))
+
+    completion = await openai.ChatCompletion.acreate(
+        model=AI_ENGINE,
+        messages=conversation.messages
+    )
+    logging.debug(f"Completion: {completion}")
+
+    response_message = completion.choices[0]['message']
+    response = response_message['content'].strip()
+    logging.debug(f"Response: {response}")
+
+    conversation.messages.append(response_message)
+
+    return response
+
+
+def parse_blocks(text: str, limit=2000):
+    tick = '`'
+    block = ""
+    current_fence = ""
+    for line in text.splitlines():
+        if len(block) + len(line) > limit - 3:
+            if block:
+                if current_fence:
+                    block += '```'
+                yield block
+                block = current_fence
+
+        block += ('\n' + line) if block else line
+
+        if line.strip().startswith(tick * 3):
+            if current_fence:
+                current_fence = ""
+            else:
+                current_fence = line
+
+    if block:
+        yield block
+
+
+async def send(thread: discord.Thread, text: str):
+    for block in parse_blocks(text):
+        await thread.send(block)
+
+
+async def send_in_channel(channel: discord.TextChannel, text: str):
+    for block in parse_blocks(text):
+        await channel.send(block)
+
```

Comment:
This is identical to `send`, so I don't think we need it.

If you want to get the typing right, use `Union[discord.Thread, discord.TextChannel]`